### DISCUSSION TITLE
How to download spacy NLP models using rye?

### DISCUSSION THREAD

**Discussion by saona-raimundo** [astral-sh/rye#1004] [Q&A] [answered]
SpaCy is a Python NLP library.
Since models are big, they use a special pipeline to download them.
A basic example is
```python
> python -m spacy download en_core_web_trf

// main.py
import spacy
nlp = spacy.load("en_core_web_trf")
import en_core_web_trf
nlp = en_core_web_trf.load()
doc = nlp("This is a sentence.")
print([(w.text, w.pos_) for w in doc])
```
from https://spacy.io/models

I have trouble running the special download pipeline of SpaCy through rye.
How would you make it work?

**Comment by goschtl** [comment, marked as answer]
Hi,

it's easy you can just use this one

`rye run python -m spacy download en_core_web_trf`

you have to make sure that pip is also installed via

`rye add pip`

HTH 
Christian

**Reply by saona-raimundo** [reply]
Thank you!

My full command history was
```
rye init --script <project-name>
cd <project-name>
rye add pip
rye add spacy
rye sync
rye run python -m spacy download en_core_web_trf
```

**Reply by limarta** [reply]
It seems that after pip installs a model, rye is not informed and thus the pyproject.toml doesn't have the downloaded model. Indeed rye proceeds to uninstall `en_core_web_trf` after a `rye sync` or `rye add x`. Is it possible to prevent rye from removing the downloaded model, say, like adding it to the pyproject.toml?

**Reply by math625f** [reply]
@limarta Did you ever manage to solve this?

**Reply by cdolek** [reply]
I just asked here: https://github.com/astral-sh/rye/discussions/1387

**Reply by drothermel** [reply]
According to the Spacy [docs](https://spacy.io/usage/models#download-pip):

>> To download a trained pipeline directly using [pip](https://pypi.python.org/pypi/pip), point pip install to the URL or local path of the wheel file or archive. Installing the wheel is usually more efficient.
```
# With external URL
pip install https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.0.0/en_core_web_sm-3.0.0-py3-none-any.whl
pip install https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.0.0/en_core_web_sm-3.0.0.tar.gz

# Using spacy info to get the external URL
pip install $(spacy info en_core_web_sm --url)

# With local file
pip install /Users/you/en_core_web_sm-3.0.0-py3-none-any.whl
pip install /Users/you/en_core_web_sm-3.0.0.tar.gz
```

So I tried the following which seems to work:
`rye add en_core_web_sm --url $(spacy info en_core_web_sm --url)`

**Reply by cdolek** [reply]
@drothermel you are the best, thank you!
